File: 0121_best_time_to_buy_stock/best_time_to_buy_and_sell_stock.py
from typing import List

class BestTimeToBuyAndSellStock(object):
    def maxProfit(self, prices: List[int]) -> int:
        min_price = float('inf')
        max_profit = 0
        buy_day = sell_day = -1
        
        
        for i in range(0, len(prices)):
            if prices[i] < min_price:
                buy_day = i
                sell_day = -1
                min_price = prices[i]
                
            if prices[i] - min_price > max_profit:
                max_profit = prices[i] - min_price       
                sell_day = i
                
        print(f"Buy at {buy_day} : {prices[buy_day]}")        
        print(f"Sell at {sell_day} : {prices[sell_day]}")        
                
        return int(max_profit) 
        
        
ob = BestTimeToBuyAndSellStock()        

nums = [7, 1, 5, 3, 6, 4]
print(nums)
answer = ob.maxProfit(nums)
print(f"Answer is {answer}")

File: 0121_best_time_to_buy_stock/BestTimeToBuyAndSellStock.php
<?php

class BestTimeToBuyAndSellStock
{
    /**
     * @param int[] $prices
     * @return int
     */
    function maxProfit1($prices)
    {
        $left = 0;
        $right = 1;
        $maxP = 0;

        while ($right < count($prices)) {
            if ($prices[$left] < $prices[$right]) {
                $profit = $prices[$right] - $prices[$left];
                $maxP = max($maxP, $profit);
            } else {
                $left = $right;
            }

            $right++;
        }

        return $maxP;
    }

    function maxProfit($prices)
    {
        $min_price = PHP_INT_MAX;
        $max_profit = 0;
        $buy_day = -1;
        $sell_day = -1;

        foreach ($prices as $index => $price) {
            if ($price < $min_price) {
                $min_price = $price;
                $buy_day = $index;
                $sell_day = -1;
            } else if ($price - $min_price > $max_profit) {
                $max_profit = $price - $min_price;
                $sell_day = $index;
            }
        }


        echo "Buy at {$buy_day} : {$prices[$buy_day]}\n";
        echo "Sell at {$sell_day} : {$prices[$sell_day]}\n";

        return $max_profit;
    }
}

$solution = new BestTimeToBuyAndSellStock();

$nums = [7, 1, 5, 3, 6, 4];
print_r($nums);
echo $solution->maxProfit($nums);
echo "\n";
